runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -28,15 +28,126 @@
 #include "chpl-comm.h"
 #include "chpl-mem.h"
 #include "chplmemtrack.h"
+#include "chplsys.h"
+#include "chpl-topo.h"
 #include "chpltypes.h"
 #include "error.h"
 
-static struct shared_heap {
+
+//
+// Information about shared heaps gotten from the comm layer.
+//
+typedef struct shared_heap {
   void* base;
   size_t size;
   size_t cur_offset;
   pthread_mutex_t alloc_lock;
-} heap; // static, will be "zero" initialized automatically
+} shared_heap_t;
+
+static shared_heap_t* heaps;
+
+static int get_num_heaps(void) {
+  static int num_heaps = -1;
+
+  if (num_heaps < 0) {
+    //
+    // Initialize: without multiple NUMA domains we definitely don't
+    // need multiple heaps.  Even with multiple NUMA domains we might
+    // not need multiple heaps, if other constraints aren't satisfied.
+    //
+    void* heap_base;
+    size_t heap_size;
+
+    chpl_comm_desired_shared_heap(&heap_base, &heap_size);
+    if (heap_base == NULL) {
+      num_heaps = 0;
+    } else {
+      num_heaps = chpl_topo_getNumNumaDomains();
+      if (num_heaps < 1) {
+        num_heaps = 1;
+      } else if (!chpl_mem_impl_alloc_localizes()) {
+        num_heaps = 1;
+      }
+    }
+  }
+
+  return num_heaps;
+}
+
+static inline int get_heap(void) {
```

Comment:
Maybe rename to get_nearest_heap.. that's a bad name, but something that indicates it's getting the heap that you have affinity to?

get_heap_with_thread_locality?

get_nearest_heap

get_heap_sorry_elliot_sucks_at_naming_things

Eh.. maybe just leave the name alone (unless you have a really good one) and add a short comment noting it'll get the heap with thread locality